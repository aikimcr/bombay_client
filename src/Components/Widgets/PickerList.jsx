import { createRef, useContext } from "react";
import PropTypes from "prop-types";

import BombayLoginContext from "../../Context/BombayLoginContext";
import useModelCollection from "../../Hooks/useModelCollection";

function PickerList({ pickModel, isOpen, collectionClass }) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [listCollection] = useModelCollection({
    CollectionClass: collectionClass,
    topRef,
    loginState,
  });

  function clickHandler(evt, model) {
    evt.preventDefault();
    pickModel(model);
  }

  if (!isOpen) return null;

  return (
    <div className="picker-component">
      <div className="list-container" ref={topRef}>
        <ul className="picker-list">
          {listCollection?.current == null
            ? ""
            : listCollection.current.map((model) => {
                const key = `mode-list-${model.get("id")}`;
                return (
                  <li
                    className="picker-item"
                    key={key}
                    data-model-id={model.get("id")}
                    onClick={(evt) => {
                      clickHandler(evt, model);
                    }}
                  >
                    {model.get("name")}
                  </li>
                );
              })}
        </ul>
      </div>
    </div>
  );
}

PickerList.propTypes = {
  pickModel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  collectionClass: PropTypes.func.isRequired,
};

export default PickerList;
