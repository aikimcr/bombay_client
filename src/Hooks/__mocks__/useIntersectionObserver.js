import { useEffect, useRef } from "react";

export class mockEntry {
  constructor(target, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  static observers = [];

  constructor(callback, options) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    this.constructor.observers.push(this);
  }

  observe(target) {}

  unobserve(target) {}

  disconnect() {}

  _fireIntersect(target, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
