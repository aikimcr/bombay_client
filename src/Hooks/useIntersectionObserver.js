import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
