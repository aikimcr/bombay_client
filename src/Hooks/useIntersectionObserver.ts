import { useEffect, useRef, RefObject } from 'react';

export interface IntersectionObserverOptions {
  root?: Element | null;
  rootMargin?: string;
  threshold?: number | number[];
}

export interface UseIntersectionObserverReturn {
  current: IntersectionObserver | null;
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
): UseIntersectionObserverReturn {
  const observer = useRef<IntersectionObserver | null>(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      if (observer.current) {
        observer.current.disconnect();
      }
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
