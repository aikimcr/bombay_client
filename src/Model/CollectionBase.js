import { getFromURLString, postToURLString, prepareURLFromArgs } from '../Network/Network';
import { loginStatus } from '../Network/Login';

import ModelBase from './ModelBase';

// Use a Javascript class rather than just a module.  This allows simpler derivation.
class CollectionBase {
    static isCollection(collection) {
        return collection instanceof this;
    }

    static buildUrl(urlOrPath, query = {}) {
        try {
            const url = new URL(urlOrPath);
            for (const key in query) {
                url.searchParams.set(key, query[key]);
            }
            return url.toString();
        } catch(err) {
            return prepareURLFromArgs(urlOrPath, query).toString();
        }
    }

    #idUrl;
    #modelClass;
    #models = [];
    #prevPage;
    #nextPage;
    #readyPromise;
    #maxModels = 1000; // Trim the model list if it gets longer than this.
    // #pages = [];

    constructor(url, options = {}) {
        this.#idUrl = this.constructor.buildUrl(url);

        this.#modelClass = options.modelClass ?? ModelBase;

        if (options.models != null) {
            const models = options.models;
            this.#prevPage = options.prevPage;
            this.#nextPage = options.nextPage;
            models.forEach(model => this.add(model));
            this.#readyPromise = Promise.resolve(this.#models);
        } else {
            this.#readyPromise = this.#fetch(this.#idUrl);
        }
    }

    idUrl() {
        return this.#idUrl;
    }

    modelClass() {
        return this.#modelClass;
    }

    models() {
        return [...this.#models];
    }

    length() {
        return this.#models.length;
    }

    ready() {
        return this.#readyPromise;
    }

    add(model) {
        const newModel = this.#modelClass.from(model);
        this.#models.push(newModel);
        return newModel;
    }

    forEach(callback) {
        return this.#models.forEach((model, index, list) => {
            return callback(model, index, list);
        });
    }

    map(callback) {
        return this.#models.map((model, index, list) => {
            return callback(model, index, list);
        });
    }

    reduce(callback, start) {
        return this.#models.reduce((memo, model, index, list) => {
            return callback(memo, model, index, list)
        }, start);
    }

    #mergeModels(data, {reset, append} = {reset: true, append: true}) {
        if (reset) {
            this.#models = [];
        }

        if (data && data.length > 0) {
            const currentIds = this.#models.map(model => model.get('id'));
            const newModels = data
                .filter(def => !currentIds.includes(def.id))
                .map(def => this.#modelClass.from(def));

            if (append) {
                this.#models = [...this.#models, ...newModels].slice(-this.#maxModels);
            } else {
                this.#models = [...newModels, ...this.#models].slice(0, this.#maxModels);
            }
        }
    }

    async #fetch(url, options) {
        const loggedIn = await loginStatus();

        if (!loggedIn) {
            this.#models = [];
            this.#prevPage = null;
            this.#nextPage = null;
            return this.#models;
        }

        const getPromise = getFromURLString(url);
        // debugger;
        // const body = await getFromURLString(url)
        const body = await getPromise
            .catch((err) => {
                if (err.status !== 404) {
                    console.error(err.status, err.message);
                }

                this.#prevPage = undefined;
                this.#nextPage = undefined;

                return Promise.resolve({
                    data: [],
                    prevPage: null,
                    nextPage: null,
                });
            });

        this.#mergeModels(body.data, options);
        this.#prevPage = body.prevPage;
        this.#nextPage = body.nextPage;
        return this.#models;
    }

    async save(modelDef) {
        const newDef = await postToURLString(this.#idUrl, modelDef);
        return this.add(newDef);
    }

    hasNextPage() {
        return !!this.#nextPage;
    }

    async fetchNextPage() {
        if (this.#nextPage) {
            await this.#fetch(this.#nextPage, {reset: false, append: true});
        }

        return this.#models;
    }

    hasPrevPage() {
        return !!this.#prevPage;
    }

    async fetchPrevPage() {
        if (this.#prevPage) {
            await this.#fetch(this.#prevPage, {reset: false, append: false});
        }

        return this.#models;
    }

    // Handy to have ana also helps with testing
    toJSON() {
        return {
            data: this.#models.map(model => model.toJSON()),
            prevPage: this.#prevPage,
            nextPage: this.#nextPage,
        };
    }

    toString() {
        return JSON.stringify(this.toJSON());
    }
};

export default CollectionBase;