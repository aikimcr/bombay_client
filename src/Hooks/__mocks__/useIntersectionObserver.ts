import { useEffect, useRef, RefObject } from 'react';

import { IntersectionObserverOptions } from '../useIntersectionObserver';
export class mockEntry {
  boundingClientRect: DOMRectReadOnly;
  intersectionRatio: number;
  intersectionRect: DOMRectReadOnly;
  isIntersecting: boolean;
  rootBounds: DOMRectReadOnly;
  target: Element;
  time: number;
  isVisible: boolean;

  constructor(target: Element, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  root: Element;
  rootMargin: string;
  thresholds: readonly number[];
  callback: IntersectionObserverCallback;
  options: IntersectionObserverOptions;
  entries: mockEntry[];

  static observers: mockObserver[] = [];

  constructor(
    callback: IntersectionObserverCallback,
    options: IntersectionObserverOptions,
  ) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    mockObserver.observers.push(this);
  }

  observe(_target: Element) {}

  unobserve(_target: Element) {}

  takeRecords() {
    return this.entries;
  }

  disconnect() {}

  _fireIntersect(target: Element, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
