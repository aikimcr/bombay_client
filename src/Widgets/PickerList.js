import { useEffect, useState, useRef, createRef } from 'react';
import PropTypes from 'prop-types';

import useIntersectionObserver from '../Hooks/useIntersectionObserver';

function PickerList(props) {
    const topRef = createRef();

    const listCollection = useRef(null);
    const observer = useIntersectionObserver(topRef, (entries, observer) => {
        entries.forEach(entry => {
            if (entry.isIntersecting) {
                setShouldPage(true);
                observer.unobserve(entry.target);
            }

            // I haven't found a case where this is actually needed, but I haven't ruled it out.
            // if (entry.isVisible) {
            // }
        });
    });

    const [shouldPage, setShouldPage] = useState(false);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        if (!props.isOpen) return;

        if (listCollection.current == null) {
            refreshCollection()
        }
    }, [props.isOpen]);

    useEffect(() => {
        if (!shouldPage) { return; }
        if (loading) { return; }

        setShouldPage(false);

        if (listCollection.current.hasNextPage()) {
            setLoading(true);

            listCollection.current.fetchNextPage()
                .then(() => {
                    setLoading(false);
                });
        }
    }, [shouldPage, loading]);

    useEffect(() => {
        if (topRef.current == null) return;
        if (loading) { return; }

        if (listCollection.current && listCollection.current.hasNextPage()) {
            const myElement = topRef.current.querySelector('li:last-child');

            if (myElement) {
                observer.current.observe(myElement);
            }
        }
    }, [loading, observer, topRef]);

    function refreshCollection() {
        setLoading(true);

        listCollection.current = new props.collectionClass();
        listCollection.current.ready()
            .then(() => {
                setLoading(false);
            });
    }

    function clickHandler(evt, model ) {
        evt.preventDefault();
        props.pickModel(model);
    }

    if (!props.isOpen) return null;

    return (
        <div className="picker-component">
            <div className="list-container" ref={topRef}>
                <ul className="picker-list">
                    {listCollection?.current == null ? '' : listCollection.current.map(model  => {
                        const key = `mode-list-${model.get('id')}`;
                        return (
                            <li
                                className='picker-item'
                                key={key}
                                data-model-id={model.get('id')}
                                onClick={evt => { clickHandler(evt, model) }}
                            >
                                {model.get('name')}
                            </li>
                        );
                    })}
                </ul>
            </div>
        </div>
    );
}

PickerList.propTypes = {
    pickModel: PropTypes.func.isRequired,
    isOpen: PropTypes.bool.isRequired,
    collectionClass: PropTypes.func.isRequired,
}

export default PickerList;