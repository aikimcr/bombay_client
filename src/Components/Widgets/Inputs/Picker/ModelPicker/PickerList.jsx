import React, { createRef, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import BombayLoginContext from '../../../../../Context/BombayLoginContext';
import { useModelCollection } from '../../../../../Hooks/useModelCollection';

export function PickerList({ pickModel, isOpen, initialCollection }) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [isMounted, setIsMounted] = useState(false);
  const [listCollection] = useState(initialCollection);
  const { refreshCollection } = useModelCollection({
    initialCollection: listCollection,
    topRef,
    isMounted,
  });

  useEffect(() => {
    setIsMounted(true);
  }, []);

  function clickHandler(evt, model) {
    evt.preventDefault();
    pickModel(model);
  }

  if (!isOpen) return null;

  return (
    <div className="picker-component" data-testid="picker-component">
      <div className="list-container" ref={topRef}>
        <ul className="picker-list">
          {!listCollection
            ? ''
            : listCollection.map((model) => {
                const key = `mode-list-${model.id}`;
                return (
                  <li
                    className="picker-item"
                    data-testid="picker-item"
                    key={key}
                    data-model-id={model.id}
                    onClick={(evt) => {
                      clickHandler(evt, model);
                    }}
                  >
                    {model.name}
                  </li>
                );
              })}
        </ul>
      </div>
    </div>
  );
}

PickerList.propTypes = {
  pickModel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  collectionClass: PropTypes.func.isRequired,
};

export default PickerList;
