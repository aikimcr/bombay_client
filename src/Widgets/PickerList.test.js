import { act, render } from '@testing-library/react';

import * as Network from '../Network/Network';
jest.mock('../Network/Network');

import useIntersectionObserver, * as mockObserver from '../Hooks/useIntersectionObserver';
jest.mock('../Hooks/useIntersectionObserver');

import ModelBase from '../model/ModelBase';
import CollectionBase from '../model/CollectionBase';

class TestModel extends ModelBase{};

class TestCollection extends CollectionBase {
    constructor(options={}) {
        super('/table1', {...options, modelClass: TestModel});
    }
}

import PickerList from './PickerList';

jest.useFakeTimers();

const testToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4';

beforeEach(() => {
    localStorage.setItem('jwttoken', testToken);
})

afterEach(() => {
    while (mockObserver.mockObserver.observers.length > 0) {
        mockObserver.mockObserver.observers.pop();
    }

    localStorage.removeItem('jwttoken');
})

it('should show the list', async () => {
    const pickModel = jest.fn();
    
    const { resolve } = Network._setupMocks();

    const  { asFragment } = render(
        <PickerList
            pickModel={pickModel}
            isOpen={true}
            collectionClass={TestCollection}
        />
    );

    const [fetchBody] = makeModels(10, {});

    await act(async () => {
        resolve(fetchBody);
    });

    expect(asFragment).toMatchSnapshot();

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    expect(Network.getFromURLString).toBeCalledTimes(1);
    expect(Network.getFromURLString).toBeCalledWith('http://localhost:2001/xyzzy/table1');
    expect(Network.getFromURLString.mock.results[0].value).resolves.toEqual(fetchBody);

    const listElements = asFragment().querySelectorAll('li.picker-item');
    expect(listElements).toHaveLength(10);
});

it('should be an empty component', async () => {
    const pickModel = jest.fn();

    const { container } = render(
        <PickerList
            pickModel={pickModel}
            isOpen={false}
            collectionClass={TestCollection}
        />
    );

    expect(container).toBeEmptyDOMElement()
});

it('should show the next page', async () => {
    const pickModel = jest.fn();

    const { resolve } = Network._setupMocks();

    const { asFragment } = render(
        <PickerList
            pickModel={pickModel}
            isOpen={true}
            collectionClass={TestCollection}
        />
    );

    const [fetchBody] = makeModels(10, {});

    await act(async () => {
        resolve(fetchBody);
    });

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    expect(Network.getFromURLString).toBeCalledTimes(1);
    expect(Network.getFromURLString).toBeCalledWith('http://localhost:2001/xyzzy/table1');
    expect(Network.getFromURLString.mock.results[0].value).resolves.toEqual(fetchBody);

    const { resolve: resolve2, reject, promise } = Network._setupMockPromise();

    const [fetchBody2] = makeModels(10, {offset: 10, limit: 10});
    const lastItem = asFragment().querySelector('li.picker-item:last-child');
    const observer = mockObserver.mockObserver.observers[0];

    await act(async () => {
        observer._fireIntersect(lastItem);
        resolve2(fetchBody2);
    });

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    expect(Network.getFromURLString).toBeCalledTimes(2);
    expect(Network.getFromURLString).toHaveBeenLastCalledWith('http://localhost:2001/xyzzy/table1?offset=10&limit=10');
    expect(Network.getFromURLString.mock.results[1].value).resolves.toEqual(fetchBody2);

    const listElements = asFragment().querySelectorAll('li.picker-item');
    expect(listElements).toHaveLength(20);
}, 300000);

it('should call the callback when an item is clicked', async () => {
    const pickModel = jest.fn();

    const { resolve } = Network._setupMocks();

    const { container, queryByText } = render(
        <PickerList
            pickModel={pickModel}
            isOpen={true}
            collectionClass={TestCollection}
        />
    );

    const [fetchBody, models] = makeModels(10, {});

    await act(async () => {
        resolve(fetchBody);
    });

    expect(pickModel).not.toBeCalled();

    const el = queryByText(models[1].get('name'));
    el.click();

    expect(pickModel).toBeCalled();
    expect(pickModel).toBeCalledWith(models[1]);
});
