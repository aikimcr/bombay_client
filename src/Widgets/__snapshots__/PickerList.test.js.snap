// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show the list 1`] = `[Function]`;
